File: framework/languages/java/e2e/SpotActorTransfer/Shared/src/main/java/systems/zlink/e2e/spotactortransfer/shared/Env.java
package systems.zlink.e2e.spotactortransfer.shared;

import java.io.Reader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Properties;

public final class Env {
    private static final Properties VALUES = new Properties();

    private Env() {
    }

    public static synchronized void configure(String... args) {
        VALUES.clear();
        String path = configPath(args);
        if (path == null) {
            return;
        }
        try (Reader reader = Files.newBufferedReader(Path.of(path))) {
            VALUES.load(reader);
        } catch (Exception error) {
            throw new IllegalStateException("Could not load SpotActorTransfer E2E config", error);
        }
    }

    public static String get(String name, String fallback) {
        String value = VALUES.getProperty(name);
        return value == null || value.isBlank() ? fallback : value;
    }

    public static String require(String name) {
        String value = get(name, "");
        if (value.isBlank()) {
            throw new IllegalStateException(name + " is required");
        }
        return value;
    }

    private static String configPath(String... args) {
        for (int index = 0; index < args.length; index++) {
            if ("--e2e-config".equals(args[index]) && index + 1 < args.length) {
                return args[index + 1];
            }
            if (args[index].startsWith("--e2e-config=")) {
                return args[index].substring(args[index].indexOf('=') + 1);
            }
        }
        return null;
    }
}

File: framework/languages/java/e2e-kotlin/RegistrationCodec/Shared/src/main/kotlin/systems/zlink/e2e/kotlin/registrationcodec/Env.kt
package systems.zlink.e2e.kotlin.registrationcodec

import java.nio.file.Files
import java.nio.file.Path
import java.util.Properties

object Env {
    private val values = linkedMapOf<String, String>()

    @Synchronized
    fun configure(args: Array<out String>) {
        values.clear()
        val path = configPath(args) ?: return
        val properties = Properties()
        Files.newInputStream(Path.of(path)).use { properties.load(it) }
        for (name in properties.stringPropertyNames()) {
            values[name] = properties.getProperty(name)
        }
    }

    fun get(name: String, fallback: String = ""): String {
        val value = values[name]
        return if (value.isNullOrBlank()) fallback else value
    }

    fun applicationArgs(args: Array<out String>): Array<String> = withoutConfig(args)

    private fun configPath(args: Array<out String>): String? {
        for (index in args.indices) {
            val argument = args[index]
            if (argument == "--e2e-config" && index + 1 < args.size) {
                return args[index + 1]
            }
            if (argument.startsWith("--e2e-config=")) {
                return argument.substringAfter('=', "")
            }
        }
        return null
    }

    private fun withoutConfig(args: Array<out String>): Array<String> {
        val result = mutableListOf<String>()
        var index = 0
        while (index < args.size) {
            val argument = args[index]
            if (argument == "--e2e-config") {
                index += 2
            } else if (argument.startsWith("--e2e-config=")) {
                index++
            } else {
                result += argument
                index++
            }
        }
        return result.toTypedArray()
    }

}
